### PR TITLE
Added support for PowerPC

### PR DESCRIPTION
**Pull request by zacharyselk**
Adds PowerPC to the known targets and updates some tests to use the system assembler rather than the GNU assembler.

### CODE REVIEW

File: flang/lib/Optimizer/CodeGen/Target.cpp [hunk 1]
```diff
@@ -205,6 +205,55 @@ struct TargetAArch64 : public GenericTarget<TargetAArch64> {
...
+    const auto *sem = &floatToSemantics(kindMap, eleTy);
+    if (sem == &llvm::APFloat::IEEEsingle()) {
+      // <2 x t>   vector of 2 eleTy
+      marshal.emplace_back(fir::VectorType::get(2, eleTy), AT{});
+    } else if (sem == &llvm::APFloat::IEEEdouble()) {
+      // two distinct double arguments
+      marshal.emplace_back(eleTy, AT{});
+      marshal.emplace_back(eleTy, AT{});
+    } else {
+      llvm_unreachable("not implemented");
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/CodeGen/Target.cpp:230]: At some point soon, we'll have to consider if/how to support quad precision as well. (But we're not _quite_ ready yet, so no worries for now.)

**Review by schweitzpgi — APPROVED** [review]
LGTM.

File: flang/lib/Optimizer/CodeGen/Target.cpp [hunk 2]
```diff
@@ -205,6 +205,55 @@ struct TargetAArch64 : public GenericTarget<TargetAArch64> {
 };
 } // namespace
 
+//===----------------------------------------------------------------------===//
+// PPC64le (PowerPC 64 bit little endean) linux target specifics.
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct TargetPPC64le : public GenericTarget<TargetAArch64> {
```
Comment (kiranchandramohan) [inline flang/lib/Optimizer/CodeGen/Target.cpp:213]: Should this be GenericTarget<TargetAArch64> or GenericTarget<TargetPPC64le>?

**Review by kiranchandramohan — COMMENTED** [review]
I saw different parameter and return types for Complex in  powerpc64le using clang while using two different test programs.

Result
```
declare { float, float } @__fc_cos_1(float, float)
declare [2 x float] @complex_swap([2 x float]) #2
```

Programs given below. Sources are the following links.
https://github.com/flang-compiler/f18-llvm-project/issues/344#issuecomment-666325822
https://llvm.discourse.group/t/c-representation-of-complex-types/2180/6
```
#include <complex.h>
#include <stdio.h>

float _Complex __fc_cos_1(float _Complex);

int main(void) {
  float _Complex z1, z2;
  z1 = 3.125 + I*1.5;
  z2 = __fc_cos_1(z1);
  printf("(%f, %f)\n", creal(z2), cimag(z2));
}
```

```
typedef struct {
  float re ;
  float im ;
} complex ;

complex complex_swap(complex);

int main() {
  complex x ;
  x.re = 4 ;
  x.im = 3 ;
  complex y = complex_swap(x) ;
  return 0 ;
}
```
1. Could you check and confirm what should be correct setting for clang for powerpc ?
2. Also can the complex.fir tests be duplicated for power and conditionally enabled for the power platform? Actually, how is this test passing on other platforms?

[Not a comment on this patch] And now that I look at it, I see that i submitted a wrong patch for AArch64. The patch was supposed to be the following. I will post a new patch after doing some tests.
```
+//===----------------------------------------------------------------------===//
+// AArch64 (AArch64 bit) linux target specifics.
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct TargetAArch64 : public GenericTarget<TargetAArch64> {
+  using GenericTarget::GenericTarget;
+
+  static constexpr int defaultWidth = 64;
+
+  CodeGenSpecifics::Marshalling
+  complexArgumentType(mlir::Type eleTy) const override {
+    CodeGenSpecifics::Marshalling marshal;
+    const auto *sem = &floatToSemantics(kindMap, eleTy);
+    if (sem == &llvm::APFloat::IEEEsingle() ||
+        sem == &llvm::APFloat::IEEEdouble()) {
+      // [2 x t]   array of 2 eleTy
+      marshal.emplace_back(fir::SequenceType::get({2}, eleTy), AT{});
+    } else {
+      llvm_unreachable("not implemented");
+    }
+    return marshal;
+  }
+
+  CodeGenSpecifics::Marshalling
+  complexReturnType(mlir::Type eleTy) const override {
+    CodeGenSpecifics::Marshalling marshal;
+    const auto *sem = &floatToSemantics(kindMap, eleTy);
+    if (sem == &llvm::APFloat::IEEEsingle() ||
+        sem == &llvm::APFloat::IEEEdouble()) {
+      // { double, double }   struct of 2 double
+      mlir::TypeRange range = {eleTy, eleTy};
+      marshal.emplace_back(mlir::TupleType::get(range, eleTy.getContext()),
+                           AT{});
+    } else {
+      llvm_unreachable("not implemented");
+    }
+    return marshal;
+  }
+};
+} // namespace
```

**Comment by schweitzpgi** [comment]
The default complex types that I get on our Power machine with clang++ are:

```c++
std::complex<double> dbl(std::complex<double> f, std::complex<double> g)
std::complex<float> sng(std::complex<float> f)
```

```llvm
define dso_local [2 x double] @_Z3dblSt7complexIdES0_([2 x double], [2 x double]) #0 {...
define dso_local [2 x float] @_Z3sngSt7complexIfE([2 x float]) #0 {...
```
These should be double-checked against the clang sources, of course.

File: flang/lib/Optimizer/CodeGen/Target.cpp (hunk 2 above)
Comment (zacharyselk) [reply flang/lib/Optimizer/CodeGen/Target.cpp]: Ah yes, my bad. Fixed that

**Review by schweitzpgi — CHANGES_REQUESTED** [review]
Needs a rebase. Have you sorted out what the correct LLVM IR types ought to be?

**Comment by schweitzpgi** [comment]
@zacharyselk - this should be in now

**Pull request closed without merging**